### PR TITLE
feat(compiler): support unsafe json parsing

### PR DESCRIPTION
**Pull request by hasanaburayyan**
Add support for using `unsafe: true` when calling `fromJson()` methods

Closes: https://github.com/winglang/wing/issues/3788

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [x] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingsdk/src/std/bool.ts [hunk 1]
```diff
@@ -12,13 +13,16 @@ export class Boolean {
...
-  public static fromJson(json: Json): boolean {
-    if (typeof json !== "boolean") {
-      throw new Error(
-        "unable to parse " + typeof json + " " + json + " as a boolean"
-      );
-    }
-    return JSON.parse(JSON.stringify(json));
+  public static fromJson(
+    json: Json,
+    validationOptions?: JsonValidationOptions
```
Comment (Chriscbr) [inline libs/wingsdk/src/std/bool.ts:18]:
```suggestion
    options?: JsonValidationOptions
```

File: libs/wingsdk/src/std/json.ts [hunk 2]
```diff
@@ -8,6 +8,14 @@ export interface JsonStringifyOptions {
   readonly indent: number;
 }
 
+/**
+ * Options for validating Json
+ */
+export interface JsonValidationOptions {
+  /** Unsafe mode to skip validation */
```
Comment (Chriscbr) [inline libs/wingsdk/src/std/json.ts:15]:
```suggestion
  /** Unsafe mode to skip validation (may lead to runtime errors) */
```

File: libs/wingsdk/src/std/number.ts [hunk 3]
```diff
@@ -9,14 +10,19 @@ export class Number {
...
-   *
    * @param json to parse number from.
    * @returns a number.
    */
-  public static fromJson(json: Json): number {
-    json;
-    throw new Error("Macro");
+  public static fromJson(
+    json: Json,
+    validationOptions?: JsonValidationOptions
```
Comment (Chriscbr) [inline libs/wingsdk/src/std/number.ts:18]:
```suggestion
    options?: JsonValidationOptions
```

File: libs/wingsdk/src/std/string.ts [hunk 4]
```diff
@@ -20,14 +21,19 @@ export class String {
...
-   *
    * @param json to create string from.
    * @returns a string.
    */
-  public static fromJson(json: Json): string {
-    json;
-    throw new Error("Macro");
+  public static fromJson(
+    json: Json,
+    validationOptions?: JsonValidationOptions
```
Comment (Chriscbr) [inline libs/wingsdk/src/std/string.ts:29]:
```suggestion
    options?: JsonValidationOptions
```

File: libs/wingsdk/src/std/struct.ts [hunk 5]
```diff
@@ -19,17 +19,21 @@ export class Struct {
...
    * Converts a Json to a Struct
    *
-   * @macro ($self$._fromJson($args$))
+   * @macro ((json, validateOptions) => ($self$._fromJson(json, validateOptions)))($args$)
    */
-  public static fromJson(json: Json): T1 {
+  public static fromJson(
+    json: Json,
+    validateOptions?: JsonValidationOptions
+  ): T1 {
```
Comment (Chriscbr) [inline libs/wingsdk/src/std/struct.ts:27]:
```suggestion
  public static fromJson(
    json: Json,
    options?: JsonValidationOptions
  ): T1 {
```

File: docs/docs/03-language-reference.md [hunk 6]
```diff
@@ -388,6 +395,14 @@ Contact.fromJson(p);
 // - field "phone" is expected to be a string, got number.
 ```
 
+Same as with primitives, it is possible to opt-out of validation using `unsafe: true`:
+```js
+let p = Json { first: "Wing", phone: 1234 };
+let x = Contact.fromJson(p, unsafe: true);
+assert(x.last.len > 0);
+// RUNTIME ERROR: Cannot read properties of undefined (reading 'length')
```
Comment (Chriscbr) [inline docs/docs/03-language-reference.md:403]: I think writing the exact error message in the language reference could lead to folks accidentally treating it as part of the public API. The message ("cannot read properties of undefined) is also confusing to anyone not familiar with JavaScript

```suggestion
assert(x.last.len > 0); // RUNTIME ERROR
```

File: libs/wingc/src/struct_schema.rs [hunk 7]
```diff
@@ -111,6 +111,11 @@ impl<'a> StructSchemaVisitor<'a> {
 
 		// close schema
 		code.close("}");
+
+		let cleaned = code.to_string().replace("\n", "").replace(" ", "");
+
+		code = CodeMaker::default();
+		code.line(cleaned);
 		code
```
Comment (Chriscbr) [inline libs/wingc/src/struct_schema.rs:119]:
```suggestion
		CodeMaker::one_line(cleaned)
```

File: libs/wingc/src/json_schema_generator.rs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+		code.add_code(self.get_struct_schema_required_fields(&struct_.env));
+
+		// close schema
+		code.close("}");
+
+		let cleaned = code.to_string().replace("\n", "").replace(" ", "");
+
+		code = CodeMaker::default();
+		code.line(cleaned);
+		code
```
Comment (Chriscbr) [inline libs/wingc/src/json_schema_generator.rs:110]:
```suggestion
		CodeMaker::one_line(cleaned)
```

File: libs/wingc/src/json_schema_generator.rs [hunk 9]
```diff
@@ -0,0 +1,112 @@
+use crate::{
+	jsify::{codemaker::CodeMaker, JSifier},
+	type_check::{symbol_env::SymbolEnv, Struct, Type, UnsafeRef},
+};
+
+pub(crate) struct JsonSchemaGenerator {}
```
Comment (Chriscbr) [inline libs/wingc/src/json_schema_generator.rs:6]: If the struct has no fields, I think this is more idiomatic
```suggestion
pub(crate) struct JsonSchemaGenerator;
```

**Review by Chriscbr — APPROVED** [review]

File: docs/docs/03-language-reference.md (hunk 6 above)
Comment (hasanaburayyan) [reply docs/docs/03-language-reference.md:403]: I agree, I was not thinking about it when copy pasta 😁

**Review by Chriscbr — APPROVED** [review]

**Pull request merged**